### PR TITLE
new CmwLight and JsonSerialiser

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
* added CmwLightSerialiser and updated benchmarks accordingly
* added JSON (de-)serialiser and updated benchmarks accordingly
* generalized IoClassSerialiser
  * split IoClassSerialiser#deserialiseObject into data tree parsing function and actual deserialisation function
  * added handling of multiple Serialisers and auto-format detection

### CODE REVIEW

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+        assertEquals(inputObject, outputObject, "TestDataClass input-output equality");
+        return nBytes;
+    }
+
+    public static void deserialiseCustom(IoSerialiser ioSerialiser, final TestDataClass pojo) {
+        deserialiseCustom(ioSerialiser, pojo, true);
+    }
+
+    public static void deserialiseCustom(IoSerialiser ioSerialiser, final TestDataClass pojo, boolean header) {
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:145]: Codacy found an issue: [Avoid really long methods.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  9
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  10
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/IoSerialiserTests.java  6
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  13
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  7
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  26
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  -7
         

Clones added
============
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/ClassFieldDescription.java  1
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  10
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  12
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  7
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  15
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  5
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  1
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/SerialiserHelper.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416&bid=19950922)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+
+        final double diffMillis = TimeUnit.NANOSECONDS.toMillis(stopTime - startTime);
+        final double byteCount = iterations * ((byteBuffer.position() / diffMillis) * 1e3);
+        LOGGER.atInfo().addArgument(SerialiserBenchmark.humanReadableByteCount((long) byteCount, true)) //
+                .addArgument(SerialiserBenchmark.humanReadableByteCount(byteBuffer.position(), true)) //
+                .addArgument(diffMillis) //
+                .log("CmwLight Serializer (custom) throughput = {}/s for {} per test run (took {} ms)");
+    }
+
+    public static void testPerformancePojo(final int iterations, final TestDataClass inputObject, TestDataClass outputObject) {
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:373]: Codacy found an issue: [Avoid reassigning parameters such as 'outputObject'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java [hunk 3]
```diff
@@ -0,0 +1,1099 @@
...
+ * Light-weight open-source implementation of a (de-)serialiser that is binary-compatible to the serialiser used by CMW,
+ * a proprietary closed-source middle-ware used in some accelerator laboratories.
+ *
+ * N.B. this implementation is intended only for performance/functionality comparison and to enable a backward compatible
+ * transition to the {@link de.gsi.dataset.serializer.spi.BinarySerialiser} implementation which is a bit more flexible,
+ * has some additional (optional) features, and a better IO performance. See the corresponding benchmarks for details;
+ *
+ * @author rstein
+ */
+public class CmwLightSerialiser implements IoSerialiser {
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java:32]: Codacy found an issue: [Avoid really long classes.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        getFieldHeader(ioSerialiser);
+        pojo.doubleNdimArray = ioSerialiser.getDoubleArray();
+
+        final WireDataFieldDescription field = getFieldHeader(ioSerialiser);
+        if (field == null) {
+            // reached the end
+            return;
+        }
+
+        if (field.getDataType().equals(de.gsi.dataset.serializer.DataType.START_MARKER)) {
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:254]: Codacy found an issue: [Unnecessary use of fully qualified name 'de.gsi.dataset.serializer.DataType.START_MARKER' due to existing import 'de.gsi.dataset.serializer.DataType'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java [hunk 5]
```diff
@@ -792,8 +925,478 @@ public void put(final Enum<?> enumeration) {
...
+        final int nElements = n >= 0 ? Math.min(n, valuesSize) : valuesSize;
+        putArraySizeDescriptor(nElements);
+        buffer.putStringArray(values, nElements);
+        updateDataEndMarker(fieldHeader);
+    }
+
+    @Override
+    public void put(final String fieldName, final String[] values, final int[] dims) {
+        final WireDataFieldDescription fieldHeader = putFieldHeader(fieldName, DataType.STRING_ARRAY);
+        final int valuesSize = values == null ? 0 : values.length;
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java:1377]: Codacy found an issue: [Avoid unused local variables such as 'valuesSize'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  26
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  13
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  10
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  9
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  7
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/IoSerialiserTests.java  6
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  -7
         

Clones added
============
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  12
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  1
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  15
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/SerialiserHelper.java  8
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  7
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  10
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  5
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/ClassFieldDescription.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416&bid=19950922)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 2 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:373]: Codacy found an issue: [Avoid reassigning parameters such as 'outputObject'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java (hunk 5 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java:1377]: Codacy found an issue: [Avoid unused local variables such as 'valuesSize'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 1 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:145]: Codacy found an issue: [Avoid really long methods.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java (hunk 3 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java:32]: Codacy found an issue: [Avoid really long classes.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 4 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:254]: Codacy found an issue: [Unnecessary use of fully qualified name 'de.gsi.dataset.serializer.DataType.START_MARKER' due to existing import 'de.gsi.dataset.serializer.DataType'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java (hunk 5 above)
Comment (RalphSteinhagen) [reply chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java:1377]: resolved with later commit

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  9
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  26
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/IoSerialiserTests.java  6
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  10
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  13
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  7
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  -7
         

Clones added
============
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  10
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  1
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  7
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  15
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  5
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  12
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/SerialiserHelper.java  8
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/ClassFieldDescription.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416&bid=19950922)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java [hunk 6]
```diff
@@ -0,0 +1,1094 @@
...
```
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java:30]: Codacy found an issue: [Avoid really long classes.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 1 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:145]: Codacy found an issue: [Avoid really long methods.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 2 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:373]: Codacy found an issue: [Avoid reassigning parameters such as 'outputObject'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java (hunk 5 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java:1377]: Codacy found an issue: [Avoid unused local variables such as 'valuesSize'.](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

File: chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java (hunk 4 above)
Comment (RalphSteinhagen) [inline chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java:254]: Codacy found an issue: [Unnecessary use of fully qualified name 'de.gsi.dataset.serializer.DataType.START_MARKER' due to existing import 'de.gsi.dataset.serializer.DataType'](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416)

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  13
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  26
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  10
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  7
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/IoSerialiserTests.java  6
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  9
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  -7
         

Clones added
============
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  12
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  15
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  1
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  7
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/ClassFieldDescription.java  1
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  5
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/SerialiserHelper.java  8
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416&bid=19950922)

**Review by wirew0rm — APPROVED** [review]
LGTM

**Review by wirew0rm — APPROVED** [review]
Just replaced some cases of buffer.reset by buffer.flip while reviewing the api

**Comment by RalphSteinhagen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  7
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/IoSerialiserTests.java  6
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  9
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  26
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  10
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  13
         

Complexity decreasing per file
==============================
+ chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  -7
         

Clones added
============
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/JsonHelper.java  5
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/BinarySerialiser.java  12
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/SerialiserHelper.java  8
- chartfx-dataset/src/test/java/de/gsi/dataset/serializer/helper/CmwLightHelper.java  10
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/WireDataFieldDescription.java  1
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/JsonSerialiser.java  7
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/CmwLightSerialiser.java  15
- chartfx-dataset/src/main/java/de/gsi/dataset/serializer/spi/ClassFieldDescription.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=6093416&bid=19950922)

**Pull request merged**